core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationLimitTests.java

```diff
@@ -0,0 +1,220 @@
+    private ClusterService clusterService;
+    private ThreadPool threadPool;
+    private TransportService transportServiceB;
+    private NettyTransport transportA;
+    private HierarchyCircuitBreakerService circuitBreakerService;
+
+    public void setUp() throws Exception {
+        super.setUp();
+        int startPort = 11000 + randomIntBetween(0, 1000);
+        int endPort = startPort + 10;
+        Settings settings = Settings.builder()
+            .put(TransportSettings.PORT.getKey(), startPort + "-" + endPort)
+            .put(HierarchyCircuitBreakerService.IN_FLIGHT_REQUESTS_CIRCUIT_BREAKER_LIMIT_SETTING.getKey(), new ByteSizeValue(0))
+            .build();
+        ClusterSettings clusterSettings = new ClusterSettings(settings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+
+        threadPool = new ThreadPool(getClass().getName());
+        ClusterName clusterName = new ClusterName(TransportReplicationLimitTests.class.getName());
+        NetworkService networkService = new NetworkService(settings);
+
+        circuitBreakerService = new HierarchyCircuitBreakerService(settings, clusterSettings);
+        transportA = new NettyTransport(settings, threadPool, networkService, BigArrays.NON_RECYCLING_INSTANCE, Version.CURRENT, new
+            NamedWriteableRegistry(), circuitBreakerService);
+        transportServiceA = new TransportService(transportA, threadPool).start();
+        transportServiceA.acceptIncomingRequests();
+        final DiscoveryNode nodeA = new DiscoveryNode("nodeA", transportServiceA.boundAddress().publishAddress(),
+            emptyMap(), emptySet(), Version.CURRENT);
+
+        NettyTransport transportB = new NettyTransport(settings, threadPool, networkService, BigArrays.NON_RECYCLING_INSTANCE, Version
+            .CURRENT, new NamedWriteableRegistry(), circuitBreakerService);
+        transportServiceB = new TransportService(transportB, threadPool).start();
+        transportServiceB.acceptIncomingRequests();
+        nodeB = new DiscoveryNode("nodeB", transportServiceA.boundAddress().publishAddress(),
+            emptyMap(), emptySet(), Version.CURRENT);
+
+        transportA.connectToNode(nodeB);
+        clusterService = null;
+
+        clusterService = new ClusterService(settings, null, clusterSettings, threadPool, clusterName);
+        clusterService.setLocalNode(nodeA);
+        clusterService.setNodeConnectionsService(new NodeConnectionsService(Settings.EMPTY, null, null) {
+            @Override
+            public void connectToAddedNodes(ClusterChangedEvent event) {
+                // skip
+            }
+
+            @Override
+            public void disconnectFromRemovedNodes(ClusterChangedEvent event) {
+                // skip
+            }
+        });
+        clusterService.setClusterStatePublisher((event, ackListener) -> {
+        });
+        clusterService.start();
+    }
+
+    public void tearDown() throws Exception {
+        super.tearDown();
```

Comment:
Can we tear down in the reverse order that it was set up? I tend to find it easier to reason through.